File: src/636_Exclusive_Time_of_Functions.py
# 636. Exclusive Time of Functions
# https://leetcode.cn/problems/exclusive-time-of-functions/solution/han-shu-de-du-zhan-shi-jian-by-leetcode-d54e2/
class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        ans = [0] * n
        st = []
        for log in logs:
            idx, tp, timestamp = log.split(':')
            idx, timestamp = int(idx), int(timestamp)
            if tp[0] == 's':
                if st:
                    ans[st[-1][0]] += timestamp - st[-1][1]
                    st[-1][1] = timestamp
                st.append([idx, timestamp])
            else:
                i, t = st.pop()
                ans[i] += timestamp - t + 1
                if st:
                    st[-1][1] = timestamp + 1
        return ans

File: src/636_Exclusive_Time_of_Functions.rs
// 636. Exclusive Time of Functions
// https://leetcode.cn/problems/exclusive-time-of-functions/
// 4 ms 100%
// 2.2 MB 100%
impl Solution {
    pub fn exclusive_time(n: i32, logs: Vec<String>) -> Vec<i32> {
        let mut ans = vec![0; n as usize];
        let mut stk: Vec<i32> = vec![0];
        let mut cur_time = -1; 
        for log in logs.iter() {
            let log: Vec<&str> = log.split(':').collect();
            let func_id = log[0].parse::<i32>().unwrap();
            let op = log[1];
            let time = log[2].parse::<i32>().unwrap();
            
            if op == "start" {
                stk.push(func_id);
                let time = time - 1;
                ans[stk[stk.len()-2] as usize] = ans[stk[stk.len()-2] as usize] + (time-cur_time) as i32;
                cur_time = time;
            } else {
                ans[stk[stk.len()-1] as usize] = ans[stk[stk.len()-1] as usize] + (time-cur_time) as i32;
                cur_time = time;
                stk.pop();
            }
        }
    
        ans
    }
}